pysph/sph/swe/basic.py

```diff
@@ -0,0 +1,1677 @@
+"""
+Basic Equations for solving shallow water problems
+#####################
+"""
+
+from pysph.sph.equation import Equation
+from pysph.sph.integrator_step import IntegratorStep
+from pysph.sph.integrator import Integrator
+from pysph.cpy.api import declare
+from pysph.sph.wc.density_correction import gj_solve
+
+from numpy import sqrt, cos, sin, zeros, pi
+import numpy as np
+
+
+class CheckForParticlesToSplit(Equation):
+    r"""Particles are tagged for splitting if the following condition is
+    satisfied:
+
+    .. math::
+        (A_i > A_max) and (h_i < h_max) and (x_min < x_i < x_max) and (y_min <
+        y_i < y_max)
+
+    References
+    ----------
+    .. [Vacondio2010] R. Vacondio, B.D. Rodgers and P.K. Stansby, "Accurate
+    particle splitting for smoothed particle hydrodynamics in shallow water
+    with shock capturing", Int. J. Numer. Meth. Fluids, 69 (2012), pp.
+    1377-1410
+
+    """
+    def __init__(self, dest, h_max=1e9, A_max=1e9, x_min=-1e9, x_max=1e9,
+                 y_min=-1e9, y_max=1e9):
+        r"""
+        Parameters
+        ----------
+        h_max : float
+            maximum smoothing length beyond which splitting is deactivated
+        A_max : float
+            maximum area beyond which splitting is activated
+        x_min : float
+            minimum distance along x-direction beyond which splitting is
+            activated
+        x_max : float
+            maximum distance along x-direction beyond which splitting is
+            deactivated
+        y_min : float
+            minimum distance along y-direction beyond which splitting is
+            activated
+        y_max : float
+            maximum distance along y-direction beyond which splitting is
+            deactivated
+
+        """
+        self.A_max = A_max
+        self.h_max = h_max
+        self.x_min = x_min
+        self.x_max = x_max
+        self.y_min = y_min
+        self.y_max = y_max
+        super(CheckForParticlesToSplit, self).__init__(dest, None)
+
+    def initialize(self, d_pa_to_split, d_idx):
+        d_pa_to_split[d_idx] = 0.0
+
+    def post_loop(self, d_idx, d_rho0, d_A, d_h, d_x, d_y, d_pa_to_split):
```

Comment:
You can do this in one step right in the initialize by setting the value to 1 if the condition satisfies and to 0 if not.  You then avoid an unnecessary loop!